compiler/back_end/cpp/header_generator_test.py

```diff
@@ -266,6 +266,98 @@ def test_rejects_empty_enum_case(self):
+    for test in [
+        '[(cpp) namespace: "basic"]\n',
+        '[(cpp) namespace: "multiple::components"]\n',
+        '[(cpp) namespace: "::absolute"]\n',
+        '[(cpp) namespace: "::fully::qualified"]\n',
+        '[(cpp) namespace: "CAN::Be::cAPITAL"]\n',
+        '[(cpp) namespace: "trailingNumbers54321"]\n',
+        '[(cpp) namespace: "containing4321numbers"]\n',
+        '[(cpp) namespace: "can_have_underscores"]\n',
+        '[(cpp) namespace: "_initial_underscore"]\n',
+        '[(cpp) namespace: "_initial::_underscore"]\n',
+        '[(cpp) namespace: "::_initial::_underscore"]\n',
+        '[(cpp) namespace: "trailing_underscore_"]\n',
+        '[(cpp) namespace: "trailing_::underscore_"]\n',
+        '[(cpp) namespace: "::trailing_::underscore_"]\n',
+        '[(cpp) namespace: " spaces "]\n',
+        '[(cpp) namespace: "with :: spaces"]\n',
+        '[(cpp) namespace: "   ::fully:: qualified :: with::spaces"]\n',
+    ]:
+      ir = _make_ir_from_emb(test)
+      self.assertEqual([], header_generator.generate_header(ir)[1])
+
+  def test_rejects_non_namespace_strings(self):
+    for test in [
+        '[(cpp) namespace: "5th::avenue"]\n',
+        '[(cpp) namespace: "can\'t::have::apostrophe"]\n',
+        '[(cpp) namespace: "cannot-have-dash"]\n',
+        '[(cpp) namespace: "no/slashes"]\n',
+        '[(cpp) namespace: "no\\\\slashes"]\n',
+        '[(cpp) namespace: "apostrophes*are*rejected"]\n',
+        '[(cpp) namespace: "avoid.dot"]\n',
+        '[(cpp) namespace: "does5+5"]\n',
+        '[(cpp) namespace: "=10"]\n',
+        '[(cpp) namespace: "?"]\n',
+        '[(cpp) namespace: "reject::spaces in::components"]\n',
+        '[(cpp) namespace: "totally::valid::but::extra         +"]\n',
+        '[(cpp) namespace: "totally::valid::but::extra         ::?"]\n',
+        '[(cpp) namespace: "< totally::valid::but::extra"]\n',
+        '[(cpp) namespace: "< ::totally::valid::but::extra"]\n',
+        '[(cpp) namespace: "::totally::valid::but::extra::"]\n',
+        '[(cpp) namespace: ":::extra::colon"]\n',
+        '[(cpp) namespace: "::extra:::colon"]\n',
+    ]:
+      ir = _make_ir_from_emb(test)
+      attr = ir.module[0].attribute[0]
+      self.assertEqual([[
+          error.error("m.emb", attr.value.source_location,
+                      'Bad namespace, does not match expected format.')
+      ]], header_generator.generate_header(ir)[1])
+
+  def test_rejects_empty_namespace(self):
+    for test in [
+        '[(cpp) namespace: ""]\n',
+        '[(cpp) namespace: "::"]\n',
+        '[(cpp) namespace: " "]\n',
+        '[(cpp) namespace: "    "]\n',
+        '[(cpp) namespace: "  ::"]\n',
+        '[(cpp) namespace: ":: "]\n',
+        '[(cpp) namespace: " ::  "]\n',
+    ]:
+      ir = _make_ir_from_emb(test)
+      attr = ir.module[0].attribute[0]
+      self.assertEqual([[
+          error.error("m.emb", attr.value.source_location,
+                      'Bad namespace, does not match expected format.')
```

Comment:
It's pretty straightforward to add a "Empty namespace is not allowed" error message if that's preferred, could check against a regex of `"\s*(?:::)?\s*"` to use the empty message instead of the "bad namespace" message.

Reply:
That makes sense! I added this to the latest patch. If emboss ever needs to support emitting to the global namespace, it should be fairly simple to change this verification to accept that in the future.